### PR TITLE
HMA-5107 Screenshot comparison on existing components

### PR DESCRIPTION
**Pull request by chrisob55**
# 📝 Description

https://jira.tools.tax.service.gov.uk/browse/HMA-5107
  
- [ ] Updated CHANGELOG
- [ ] Updated README

# 🛠 Testing Notes

# 📸 Screenshots

### CODE REVIEW

File: sample/build.gradle [hunk 1]
```diff
@@ -4,41 +4,64 @@ apply plugin: 'kotlin-android-extensions'
...
+
     lintOptions {
         lintConfig file('lint/config.xml')
         baseline file("lint/baseline.xml")
         warningsAsErrors true
         abortOnError true
     }
+
+    packagingOptions {
+        exclude 'META-INF/ASL2.0'
```
Comment (JamieCruwys) [inline sample/build.gradle:57]: Optional: We could do `exclude 'META-INF/*` to cover ourselves from other issues.

File: sample/build.gradle [hunk 2]
```diff
@@ -4,41 +4,64 @@ apply plugin: 'kotlin-android-extensions'
...
+apply from: "./screenshots.gradle"
+
 detekt {
     config = files("detekt/config.yml")
     baseline = file("detekt/baseline.xml")
 }
 
 android {
+    compileSdkVersion compile_sdk_version
+    buildToolsVersion "30.0.3"
```
Comment (JamieCruwys) [inline sample/build.gradle:16]: Could this be extracted out so all modules use the same build tools version?

File: sample/build.gradle [hunk 3]
```diff
@@ -57,4 +80,11 @@ dependencies {
 
     // Kotlin
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+
+    androidTestImplementation 'androidx.test.espresso:espresso-contrib:3.4.0'
```
Comment (JamieCruwys) [inline sample/build.gradle:84]: These can be extracted

File: sample/screenshots.gradle [hunk 4]
```diff
@@ -17,7 +17,9 @@ ext {
 }
 
 shot {
+    tolerance = 0.05
     runInstrumentation = Boolean.valueOf(getArgument("runInstrumentation", "true"))
+    showOnlyFailingTestsInReports = true
```
Comment (JamieCruwys) [inline sample/screenshots.gradle:22]: Will we still be able to get hold of screenshots that pass? I assume so, just wanna double check

File: sample/src/androidTest/java/uk/gov/hmrc/components/sample/ScreenshotTestExtensions.kt [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+private val RecyclerView.contentHeight: Int
+    get() {
+        val adapter = adapter!!
+        val size = adapter.itemCount
+        var contentHeight = 0
+        val recyclerWidth = width
+        InstrumentationRegistry.getInstrumentation()
+            .runOnMainSync {
+
```
Comment (JamieCruwys) [inline sample/src/androidTest/java/uk/gov/hmrc/components/sample/ScreenshotTestExtensions.kt:90]:
```suggestion
```

File: sample/src/androidTest/java/uk/gov/hmrc/components/sample/TabBarTest.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+package uk.gov.hmrc.components.sample
+
+import org.junit.Test
+import uk.gov.hmrc.components.sample.rules.FontScale
+import uk.gov.hmrc.components.sample.rules.NightMode
+
+class TabBarTest : BaseActivityTest() {
+
+    @Test
+    fun screenshots() {
```
Comment (JamieCruwys) [inline sample/src/androidTest/java/uk/gov/hmrc/components/sample/TabBarTest.kt:25]: Should these functions start with `override`? The definitions in `BaseActivityTest` are marked as open

**Review by JamieCruwys — CHANGES_REQUESTED** [review]
Looks great - got a few small comments

File: sample/screenshots.gradle (hunk 4 above)
Comment (chrisob55) [reply sample/screenshots.gradle:22]: We'll have the screenshots we need in the baseline folder

File: sample/src/androidTest/java/uk/gov/hmrc/components/sample/TabBarTest.kt (hunk 6 above)
Comment (chrisob55) [reply sample/src/androidTest/java/uk/gov/hmrc/components/sample/TabBarTest.kt:25]: The tab bar test is slightly different to the rest so doesn't override any functions

File: sample/build.gradle (hunk 1 above)
Comment (chrisob55) [reply sample/build.gradle]: Addressed in 694d064736478f073fa510eb663c87a79120f863

File: sample/build.gradle (hunk 2 above)
Comment (chrisob55) [reply sample/build.gradle]: Addressed in 694d064736478f073fa510eb663c87a79120f863

File: sample/build.gradle (hunk 3 above)
Comment (chrisob55) [reply sample/build.gradle]: Addressed in 694d064736478f073fa510eb663c87a79120f863

File: sample/src/androidTest/java/uk/gov/hmrc/components/sample/ScreenshotTestExtensions.kt (hunk 5 above)
Comment (chrisob55) [reply sample/src/androidTest/java/uk/gov/hmrc/components/sample/ScreenshotTestExtensions.kt]: Addressed in 694d064736478f073fa510eb663c87a79120f863

**Pull request merged**